core/src/main/java/org/apache/iceberg/SnapshotManager.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import com.google.common.base.Preconditions;
+import java.util.HashSet;
+import java.util.Set;
+import org.apache.iceberg.exceptions.DuplicateWAPCommitException;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.util.SnapshotUtil;
+
+public class SnapshotManager extends MergingSnapshotProducer<ManageSnapshots> implements ManageSnapshots {
+
+  private SnapshotManagerOperation operation;
+  private TableMetadata base;
+  private Long targetSnapshotId = null;
+  private Set<Long> snapshotsAlreadyCherrypicked;
+
+  enum SnapshotManagerOperation {
+    CHERRYPICK,
+    ROLLBACK
+  }
+
+  SnapshotManager(TableOperations ops) {
+    super(ops);
+    this.base = ops.current();
+    this.snapshotsAlreadyCherrypicked = new HashSet<>();
+  }
+
+  @Override
+  protected ManageSnapshots self() {
+    return this;
+  }
+
+  @Override
+  protected String operation() {
+    Snapshot manageSnapshot = base.snapshot(targetSnapshotId);
+    return manageSnapshot.operation();
+  }
+
+  @Override
+  public ManageSnapshots cherrypick(long snapshotId) {
+    Preconditions.checkArgument(base.snapshot(snapshotId) != null,
+        "Cannot cherry pick unknown snapshot id: %s", snapshotId);
+
+    operation = SnapshotManagerOperation.CHERRYPICK;
+    this.targetSnapshotId = snapshotId;
+
+    // Pick modifications from the snapshot
+    Snapshot cherryPickSnapshot = base.snapshot(this.targetSnapshotId);
+    String wapId = stagedWapId(cherryPickSnapshot);
+    boolean isWapWorkflow =  wapId != null && !"".equals(wapId);
+    if (isWapWorkflow) {
+      if (base.isWapIdPublished(wapId)) {
```

Comment:
I don't think that TableMetadata should expose convenience methods like this. Can we use a separate util class for WAP instead? That will keep the API smaller and more focused.

Reply:
+1

Reply:
I'v moved `isWapIdPublished()` to a `WapUtil` class (which is called by SnapshotManager). Although, we want to keep the complexity of the check to O(1), so a `snapshotsByWapId` (or equivalent) index would need to be built once for each table state. Additionally, Util classes shouldn't be keeping state, i'v exposed `snapshotsByWapId()` as is in  `TableMetadata` which is used by WapUtil from within the convenience method: `isWapIdPublished`.  So the state is maintained by TableMeta but the explicit check is performed by WapUtil.. Lemme know what you think.